File: Python/186__Reverse_Words_in_a_String_II.py
class Solution(object):
    def reverseWords(self, str):
        """
        :type str: List[str]
        :rtype: void Do not return anything, modify str in-place instead.
        """
        def reverse(start, end):
            while start < end:
                str[start], str[end] = str[end], str[start]
                start += 1
                end -= 1

        reverse(0, len(str) - 1)

        index, left = 0, 0
        while index < len(str):
            if str[index] == " ":
                reverse(left, index - 1)
                left = index + 1
            elif index == len(str) - 1:
                reverse(left, index)
            index += 1

File: Java/186__Reverse_Words_in_a_String_II.java
class Solution {
    public void reverseWords(char[] str) {
        reverse(str, 0, str.length - 1);
        int i = 0;
        while (i < str.length) {
            int j = i;
            while (j < str.length && str[j] != ' ') j++;
            reverse(str, i, j - 1);
            i = j;
            while (i < str.length && str[i] == ' ') i++;
        }
    }

    private void reverse(char[]str, int i, int j) {
        while (i <= j) {
            char c = str[i];
            str[i] = str[j];
            str[j] = c;
            i++;
            j--;
        }
    }
}
